File: LeetCode/40. Combination Sum II.cpp
// Author : @AdarshGaur

class Solution {
public:
    
    void backtrack(vector<int>& candidates, vector<vector<int>>& ans, vector<int>& curState, int& left, int indx){
        if(left < 0){
            return ;
        }else if(indx == candidates.size()){
            if(left == 0)
                ans.push_back(curState);
            return ;
        }else if(!curState.empty() and candidates[indx] != curState.back() and left == 0){
            ans.push_back(curState);
            return ;
        }
        
        if(curState.empty() or curState.back() != candidates[indx])
            backtrack(candidates, ans, curState, left, indx+1);
        
        left -= candidates[indx];
        curState.push_back(candidates[indx]);
        backtrack(candidates, ans, curState, left, indx+1);
        curState.pop_back();
        left += candidates[indx];
        
        return ;
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> curState;
        backtrack(candidates, ans, curState, target, 0);
        return ans;
    }
};

File: LeetCode/40. Combination Sum II.py
#Author: @AdarshGaur

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []

        def backtrack(candidates, curState, indx, target):
            if target == 0:
                ans.append(curState)
                return
            elif target < 0 or indx == len(candidates):
                return
            
            for i in range(indx, len(candidates)):
                if i > indx and candidates[i] == candidates[i-1]: continue
                backtrack(candidates, curState + [candidates[i]], i +1, target - candidates[i])
            
            return ans


        candidates.sort()
        backtrack(candidates, [], 0, target)
        return ans
